[SYCL] Implement basic reduction for parallel_for() accepting nd_range

v-klochkov opened the pull request:
This patch adds the algorithm that implements 1 reduction in parallel_for().
It handles all types and operations, including user's custom ones.
The more efficient variants are on the way.

What is NOT supported by this patch:
- parallel_for(range, ...) // i.e. simple range without work-group sizes
- parallel_for(nd_range, reduction1, reduction1, ...) // i.e. more than
  1 reductions in paralell_for
- USM
- vector reductions (dims > 1 & #elements > 1)
- HOST. The implementation used in this patch uses barrier(), which
  is not supported on HOST yet.

v-klochkov commented:
Please see the current proposal for reduction feature here: https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/Reduction/Reduction.md

The first patch adding reduction.hpp file and reduction classes is here:
https://github.com/intel/llvm/pull/1585

AlexeySachkov commented on sycl/include/CL/sycl/detail/cg.hpp:
```diff
@@ -46,7 +46,7 @@ class interop_handler {
 public:
   using QueueImplPtr = std::shared_ptr<detail::queue_impl>;
-  using ReqToMem = std::pair<detail::Requirement*, pi_mem>;
+  using ReqToMem = std::pair<detail::Requirement *, pi_mem>;
```
This seems to be an unrelated change

v-klochkov replied:
It was done by clang-format

AlexeySachkov commented on sycl/include/CL/sycl/handler.hpp:
```diff
@@ -190,14 +227,24 @@ class __SYCL_EXPORT handler {
+  void saveCodeLoc(detail::code_location CodeLoc) { MCodeLoc = CodeLoc; }
+
+  /// Stores the given \param Event to the \param Queue.
+  /// Even thought MQueue is a field of handler, the method addEvent() of
```
```suggestion
  /// Even though MQueue is a field of handler, the method addEvent() of
```

?

v-klochkov replied:
Fixed. Thank you.

AlexeySachkov commented on sycl/test/reduction/reduction_nd_conditional.cpp:
```diff
@@ -0,0 +1,131 @@
+template <typename T, int Dim, class BinaryOperation>
+class Unknown;
+
+template <typename T>
```
You could create a header file for definition of this class, since it is used in several tests

Pennycook replied:
Agreed.  I'd be tempted to rename it something like `CustomType` or `CustomVec` as well, to highlight why you're defining a new class instead of using `sycl::vec`.

v-klochkov replied:
Ok, I created a new reduction_utils.hpp header and renamed the type. Thank you.

capatober commented on sycl/include/CL/sycl/handler.hpp:
```diff
@@ -729,6 +810,305 @@ class __SYCL_EXPORT handler {
+    auto In = Redu.getReadAccToPreviousPartialReds(*this);
+    auto Out = Redu.getWriteAccForPartialReds(NWorkGroups, KernelRun, *this);
+
+    if (!InefficientCase) {
```
May `if(EfficientCase)` be better than double negation from your point of view?

v-klochkov replied:
I thought about that too. If it causes questions, it makes sense to fix it. Thank you.

capatober commented on sycl/include/CL/sycl/handler.hpp:
```diff
@@ -729,6 +810,305 @@ class __SYCL_EXPORT handler {
+        size_t GID = NDIt.get_global_linear_id();
+        // Copy the element to local memory to prepare it for tree-reduction
+        LocalReds[LID] = (GID < NWorkItems) ? In[GID] : ReduIdentity;
+        LocalReds[WGSize] = ReduIdentity;
```
Will it be better to allow only one `NDIt` write to `WGSize` to not have multiple write operations from different items?

v-klochkov replied:
Excuse me, I do not understand what you suggest here. Please give more details.

capatober replied:
I mean that you have `WGSize` work items: `0,1,..., WGSize-1`. Each of work-items updates `LocalReds[WGSize]`. Would it better to have something like that `if(LID==0){LocalReds[WGSize] = ...}`. So, only one work item (not all) updates the value.

v-klochkov replied:
If you don't mind I'll try your suggestion in the next patch.
It is a tiny fix that should not hold the commit, because it will not change performance in any way.

Having additional conditional in the code is also bad for performance. (Hmm, both the conditional and that duplicated write could be eliminated by vectorizer that combines ops across work-items), 
I'll check the device code and do or not do the additional fix with the next patch which is currently blocked by this PR.

capatober replied:
OK. Thanks for the response

capatober commented on sycl/include/CL/sycl/handler.hpp:
```diff
@@ -729,6 +810,305 @@ class __SYCL_EXPORT handler {
+          PrevStep = CurStep;
+        }
+
+        // Compute the the partial sum/reduction for the work-group.
```
```suggestion
        // Compute the partial sum/reduction for the work-group.
```

v-klochkov replied:
Fixed. Thank you.

Pennycook commented on sycl/include/CL/sycl/handler.hpp:
```diff
@@ -729,6 +810,305 @@ class __SYCL_EXPORT handler {
+  /// user's lambda function \param KernelFunc and does one iteration of
+  /// reduction of elements in each of work-groups.
+  /// This version uses tree-reduction algorithm to reduce elements in each
+  /// of work-groups. At the end of each work-groups the partial sum is written
```
```suggestion
  /// of work-groups. At the end of each work-group the partial sum is written
```

v-klochkov replied:
Fixed. Thank you.

Pennycook commented on sycl/include/CL/sycl/handler.hpp:
```diff
@@ -729,6 +810,305 @@ class __SYCL_EXPORT handler {
+    // size may be not power of those. Those two cases considered inefficient
+    // as they require additional code and checks in the kernel.
+    bool IsUnderLoaded = NWorkGroups * WGSize != NWorkItems;
+    size_t InefficientCase = (IsUnderLoaded || (WGSize & (WGSize - 1))) ? 1 : 0;
```
It took me a while to work out why this is declared as a `size_t`.

Do you think it would be clearer if `InefficientCase` was a `bool`?  You could shift the logic of whether you need to add 1 or 0 to line 945.

v-klochkov replied:
Yes, I did that. Thank you.

Pennycook commented on sycl/include/CL/sycl/handler.hpp:
```diff
@@ -729,6 +810,305 @@ class __SYCL_EXPORT handler {
+
+    // Use local memory to reduce elements in work-groups into 0-th element.
+    // If WGSize is not power of two, then WGSize+1 elements are allocated.
+    // The additional last element is used to catch reduce elements that
```
```suggestion
    // The additional last element is used to catch elements that
```

v-klochkov replied:
Fixed.

Pennycook commented on sycl/include/CL/sycl/handler.hpp:
```diff
@@ -729,6 +810,305 @@ class __SYCL_EXPORT handler {
+      handler AuxHandler(QueueCopy, MIsHost);
+      AuxHandler.saveCodeLoc(MCodeLoc);
+
+      // The last kernel DOES write to reductions's accessor.
```
```suggestion
      // The last kernel DOES write to reduction's accessor.
```

v-klochkov replied:
Fixed.

Pennycook commented on sycl/source/handler.cpp:
```diff
@@ -13,12 +13,24 @@
+}
+
+event handler::finalize() {
+  // This block of code is needed only to 5th/default reduction implementation.
```
Assuming that what you've pushed here is the default implementation:
```suggestion
  // This block of code is needed only for default reduction implementation.
```

v-klochkov replied:
Yes. Fixed.

Pennycook commented on sycl/test/reduction/reduction_nd_s0_dw.cpp:
```diff
@@ -0,0 +1,149 @@
+// RUN: %CPU_RUN_PLACEHOLDER %t.out
+// RUN: %GPU_RUN_PLACEHOLDER %t.out
+// RUN: %ACC_RUN_PLACEHOLDER %t.out
+//==----------------reduction_ctor.cpp - SYCL reduction basic test ---------==//
```
```suggestion
//==----------------reduction_nd_s0_dw.cpp - SYCL reduction basic test ---------==//
```

Although I think @bader said before that we can drop these licenses from tests, if we want to.

v-klochkov replied:
That is something new (I see that new sycl/test/abi/* tests don't have it. I dropped licenses.

v-klochkov commented:
Thank you for quick response/review. 
I added the additional changes as a separate (5th) commit of this PR.
Please re-review.

v-klochkov commented:
@Pennycook 
Ops, I missed one more comment. This one. It will take some time for me to add another test case.


> Could you please add a few tests using transparent functors as in https://github.com/intel/llvm/blob/sycl/sycl/test/group-algorithm/reduce.cpp#L75? I don't think this needs to be exhaustive, so just adding variants to the "Check with various operations" tests would be good.
> 
> Do you think it's clear what an "Aux" kernel is? It's clear to me, but may not be clear to readers less familiar with reductions. I don't want this point to block the merge, but thought I'd bring it up in case you had any ideas for alternative names.
>

v-klochkov commented:
> @Pennycook
> Ops, I missed one more comment.

This time I added a test to check reductions using transparent operators. Please see the 6th commit.

Regarding 'Aux' and naming. I don't see very good alternatives right now. Perhaps the comments, I added before reduAuxCGFunc() and inside inside parallel_for() lowering, help understand what they do.

tovinkere commented on sycl/source/handler.cpp:
```diff
@@ -13,12 +13,24 @@
+  Queue->addEvent(std::move(Event));
+}
+
+event handler::finalize() {
```
Ordered_queue sends in the code location information through finalize as a parameter. This change will affect ordered_queue. Please ensure that ordered_queue is correct as well.

v-klochkov replied:
Would you please show the code that calls finalize() method and passes code_location to it?
I grepped all files in SYCL folder and did not find such calls.
The only finalize(code_loc) call was in queue_impl::submit_impl(), which is used by ordered_queue.

But I fixed submit_impl, so no additional changes required.

tovinkere replied:
@v-klochkov  I looked at it too and the share the queue_impl, so everything looks good.

romanovvlad commented on sycl/include/CL/sycl/handler.hpp:
```diff
@@ -729,6 +810,307 @@ class __SYCL_EXPORT handler {
+      addEventToQueue(QueueCopy, MLastEvent);
+
+      // TODO: here the work-group size is not limited by user's needs,
+      // the better strategy here is to make the work-group-size as big
```
Why this is the better strategy?

v-klochkov replied:
Because user may specify nd_range with small work-group because of some good reasons requirements for user's lambda, for example nd_range=(1M, 4), So, the main kernel must obey that and use WGSize=4, but it is obviously more efficient to use bigger WGSize for aux kernels that only reduce elements and do that much faster when WGSize is big (it also will require less calls of aux kernel, when the WGSize is bigger).

I'll update comment.

romanovvlad commented on sycl/include/CL/sycl/handler.hpp:
```diff
@@ -729,6 +810,307 @@ class __SYCL_EXPORT handler {
+    // correct.
+    if (NWorkGroups > 1)
+      dissociateWithHandler(Redu.MAcc);
+
```
Suggest creating separate handler even to run user's lambda. And set "command type" of this handler to something like a NOP or Aggregator so finalize does nothing for such a type.

v-klochkov replied:
What is the reasoning for that?
That approach seems more error prone. User may create many accessors, which get associated with 'this' handler and not with the new handler. New handler will not have any knowledge about those objects
User may do some additional calls for 'this' handler which will not have effect to new handler running user's code.

v-klochkov commented on sycl/include/CL/sycl/handler.hpp:
```diff
@@ -729,6 +810,307 @@ class __SYCL_EXPORT handler {
+    // The additional last element is used to catch elements that could
+    // otherwise be lost in the tree-reduction algorithm.
+    size_t NumLocalElements = WGSize + (IsEfficientCase ? 0 : 1);
+    auto LocalReds = Redu.getReadWriteLocalAcc(NumLocalElements, *this);
```
That is one of reasons why I have that func in reduction_impl class.
Because the buffer.hpp and accessor.hpp cannot be included in handler.hpp.
I'll change if you really insist/repeat the request, I would prefer this 'auto'.

v-klochkov commented on sycl/include/CL/sycl/handler.hpp:
```diff
@@ -729,6 +810,307 @@ class __SYCL_EXPORT handler {
+  ///
+  /// Briefly: user's lambda, tree-reduction, CUSTOM types/ops.
+  template <typename KernelName, typename KernelType, int Dims, class Reduction>
+  void reduCGFunc(KernelType KernelFunc, const nd_range<Dims> &Range,
```
I'll try that.

v-klochkov replied:
I'll move that to sycl/intel/reduction.hpp and make them separate routines. I don't see any value in adding them as static method to 'reducer' or 'reduction_impl' class.

v-klochkov commented on sycl/include/CL/sycl/handler.hpp:
```diff
@@ -729,6 +810,307 @@ class __SYCL_EXPORT handler {
+    while (NWorkItems > 1) {
+      // Before creating another kernel, add the event from the previous kernel
+      // to queue.
+      addEventToQueue(QueueCopy, MLastEvent);
```
It is done to mimic the dependencies chain that is created for normal SYCL code without reductions. I.e.:
Q.submit( { user's CGFunc and lambda func})
while (NWorkGroups > 1)
  Q.submit( { Aux function to reduce elements; })

v-klochkov commented on sycl/include/CL/sycl/handler.hpp:
```diff
@@ -729,6 +810,307 @@ class __SYCL_EXPORT handler {
+    // 1. Call the kernel that includes user's lambda function.
+    // If this kernel is going to be now last one, i.e. it does not write
+    // to user's accessor, then detach user's accessor from this kernel
+    // to make the dependencies between accessors and kernels more clean and
```
There will be redundant dependencies. Probably, nothing incorrect.
Let suppose there are
1) initialization of user's reduction accessor
2) main CGFunc (with user's lambda and 1 iteration to reduce elements)
3) Aux kernel 1
4) Aux kernel 2
5) Final Aux kernel 3

If do not detach reduction accessor:                              If detach reduction accessor:
```
(1)                                                                (2) can be started even before (1)
(2) depends on (1)                                       (3) depends on (2)
(3) depends on (2)                                       (4) depends on (3)
(4) depends on (3)                                       (1) may init reduction's buffer here 
(5) depends on (4) and (1)                           (5) depends on (5) and (1)
```

So, do you think I need just to remove dissociateWithHandler?

v-klochkov commented on sycl/include/CL/sycl/handler.hpp:
```diff
@@ -729,6 +810,307 @@ class __SYCL_EXPORT handler {
+    // to user's accessor, then detach user's accessor from this kernel
+    // to make the dependencies between accessors and kernels more clean and
+    // correct.
+    if (NWorkGroups > 1)
```
If the mainCGFunc is the last kernel, i.e. if nd_range was (N, N), then mainCGFunc and the kernel in it will write to user's accessor. 
Otherwise (if NWorkGroups > 1, i.e. number of partial sums > 1), there will be another kernel following to the main kernel. and thus main kernel does not write to user's reduction variable and the dependency to user's variable is redundant.

v-klochkov commented on sycl/include/CL/sycl/handler.hpp:
```diff
@@ -729,6 +810,307 @@ class __SYCL_EXPORT handler {
+    bool IsUnderLoaded = (NWorkGroups * WGSize - NWorkItems) != 0;
+    bool IsEfficientCase = !IsUnderLoaded && ((WGSize & (WGSize - 1)) == 0);
+
+    bool IsUpdateOfUserAcc =
```
Please explain what it changes.

IsUpdateOfUserAcc check includes 'NWorkGroups == 1', which is not known statically, and if so,then how this case is different from all other temp variables such as IsEfficientCase, etc.
Do you ask using 'const TYPE Var = <something>;'  when 'Var' is initialized once and not changed after?
